src/main/kotlin/org/veupathdb/service/eda/compute/EDA.kt

```diff
@@ -0,0 +1,182 @@
+   *
+   * @param err Optional exception provider that will be used to get the
+   * exception that will be thrown if the target study does not exist.
+   *
+   * @return The `APIStudyDetail` information returned from the EDA Subsetting
+   * Service.
+   *
+   * @throws Exception Throws the exception returned by the given exception
+   * provider ([err]).
+   */
+  @JvmStatic
+  @JvmOverloads
+  fun requireAPIStudyDetail(
+    studyID: String,
+    auth: TwoTuple<String, String>,
+    err: (studyID: String) -> Exception = this::noStudyDetail
+  ): APIStudyDetail = getAPIStudyDetail(studyID, auth).orElseThrow { err(studyID) }
+
+  /**
+   * Fetches tabular study data from the EDA Merge Service for the given params.
+   *
+   * @param refMeta TODO: What is this?
+   *
+   * @param filters TODO: What is this?
+   *
+   * @param spec TODO: What is this?
+   *
+   * @param auth Auth header sent in with the job HTTP request.
+   *
+   * @return An [InputStream] over the tabular data returned from the EDA Merge
+   * Service.
+   */
+  @JvmStatic
+  fun getMergeData(
+    refMeta: ReferenceMetadata,
+    filters: List<APIFilter>,
+    spec: StreamSpec,
+    auth: TwoTuple<String, String>
+  ): InputStream =
+    EdaMergingClient(ServiceOptions.edaMergeHost, auth)
+      .getTabularDataStream(refMeta, filters, spec).inputStream
+
+
+  /**
+   * Submits a new compute job to the queue.
+   *
+   * @param plugin Plugin details.
+   *
+   * @param payload HTTP request containing the compute config.
+   *
+   * @param auth: Auth header sent in with the job HTTP request.
+   *
+   * @return A response describing the job that was created.
+   */
+  @JvmStatic
+  fun <R : ComputeRequestBase> getOrSubmitComputeJob(
+    plugin:  PluginMeta<R>,
+    payload: R,
+    auth:    TwoTuple<String, String>,
+  ): JobResponse {
+    // Serialize the http request to json
+    val serial = Json.convert(payload)
+
+    val jobID = JobIDs.of(plugin.urlSegment, serial)
+
+    // If the job already exists, just return it.
+    AsyncPlatform.getJob(jobID)?.let { return it.toJobResponse() }
+
+    Log.info("Submitting job {} to the queue", jobID)
+
+    // Build the rabbitmq message payload
+    val jobPay = PluginJobPayload(plugin.urlSegment, serial, auth.toAuthTuple())
+
+    // Submit the job
+    AsyncPlatform.submitJob(plugin.targetQueue.queueName) {
+      this.jobID  = jobID
+      this.config = Json.convert(jobPay)
+    }
+
+    // Look up the job we just submitted
+    return AsyncPlatform.getJob(jobID)!!.toJobResponse()
+  }
+
+  @JvmStatic
+  fun getComputeJobFiles(plugin: PluginMeta<out ComputeRequestBase>, payload: ComputeRequestBase): List<JobFileReference> {
+    val jobID = JobIDs.of(plugin.urlSegment, payload)
+
+    // Get the target job (or throw 404) and ensure that it is finished (or
+    // throw 403)
+    (AsyncPlatform.getJob(jobID) ?: throw NotFoundException())
+      .status
+      .isFinished || throw ForbiddenException()
```

Comment:
Is 403 the right status code if the job is in progress as opposed to an even 400? I'm also wondering if we need an API to check on a job's status. After submitting a job, how will the client know it's ready for them to retrieve the results?

Reply:
I used 403 as the operation is "Forbidden" in that it is disallowed.

It's not 100% clear to me if that is a misuse of the code or not, but I believe this is a correct interpretation of the code's intent.

The [original RFC](https://datatracker.ietf.org/doc/html/rfc2616#section-10.4.4) says:

>  The server understood the request, but is refusing to fulfill it.  Authorization will not help and the request SHOULD NOT be repeated.

The [newer RFC](https://datatracker.ietf.org/doc/html/rfc7231#section-6.5.3) says:

> The 403 (Forbidden) status code indicates that the server understood the request but refuses to authorize it.
...
a request might be forbidden for reasons unrelated to the credentials.

Reply:
Also, there are 2 ways to use the API to test the job status:

- Resubmitting a job with an identical configuration will return the existing job's status.
- Using the job ID with the JobsController GET endpoint.